tests/ui/search_is_some.stderr

```diff
@@ -35,8 +35,53 @@ LL | |                    ).is_some();
    |
    = help: this is more succinctly expressed by calling `any()`
 
+error: called `is_some()` after searching an `Iterator` with `find`
+  --> $DIR/search_is_some.rs:41:20
+   |
+LL |     let _ = (0..1).find(|x| **y == *x).is_some(); // one dereference less
+   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+help: use `any()` instead
+   |
+LL |     let _ = (0..1).any(|x| **y == *x); // one dereference less
+   |                    ^^^^^^^^^^^^^^^^^^
+help: ...and remove deref
+   |
+LL |     let _ = (0..1).find(|x| **y == x).is_some(); // one dereference less
+   |                                    ^
```

Comment:
to be honest, I wonder if we can do better here, because the suggestion is made over the `find(..).is_some()` original span, and I find it quite misleading.

Reply:
I only took a quick look, so please correct me if this is wrong. Currently, the suggestion is created by snipping the search expression and formatting that together with the `any()` search. The code then iterates over each reference and marks them in an additional help message to the user. All `*x` should be replaced with a simple `x` right?

If this assumption is correct, then we can maybe do better by doing some [`Span`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html)-Magic :sparkles:. My idea would be to not snip the entire search closure at once, but go through it step by step like this:

1. First get the closure and create a visitor similar to what you've already done.
2. Find references to `x`
3. Snip code until reference of `x`. You can use [`Span::until()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html#method.until) to get the `Span` for that
4. Continue with the search from `x` until the next reference or the end of the closure
5. Combine all collected valid code snippets

This might be a bit confusing, here is an illustration:

```rs
//       The initial span of the closure
//       vvvvvvvvvvvvvvvvvvvvvv
    find(|x| *x == y && *x == z).is_some()
//           ^^ Visitor finds reference of `x`
//       ^^^^ Get valid code pice "|x| "
//
//            v Continue search from here
    find(|x| *x == y && *x == z).is_some()
//                      ^^ Visitor finds reference of `x`
//            ^^^^^^^^^^ Get valid code pice "x == y && "
//
//                       v Continue search from here
    find(|x| *x == y && *x == z).is_some()
//                             | End of closure
//                       ^^^^^^ Get valid code pice "x == z"
//
// Combine all code snippets to this: "|x| x == y && x == z"
```

Did I understand you correctly that you were looking for a solution like this? :upside_down_face:

Reply:
Nice one, I totally missed that `Span` is this powerful :wink: